e2e/testcases/reconciler_finalizer_test.go

```diff
@@ -0,0 +1,702 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/metadata"
+	"kpt.dev/configsync/pkg/reconcilermanager"
+	"kpt.dev/configsync/pkg/status"
+	"kpt.dev/configsync/pkg/testing/fake"
+	"sigs.k8s.io/cli-utils/pkg/common"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// TestReconcilerFinalizer_Orphan tests that the reconciler's finalizer
+// correctly handles Orphan deletion propegation.
+func TestReconcilerFinalizer_Orphan(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.MultiRepos, ntopts.SkipMonoRepo)
+
+	rootRepo := nt.RootRepos[configsync.RootSyncName]
+	deployment1Name := "helloworld-1"
+
+	nt.T.Cleanup(func() {
+		// Stop webhook to avoid deletion prevention.
+		// Webhook will be re-enabled by test setup, if the next test needs it.
+		nomostest.StopWebhook(nt)
+
+		// Delete RootSync
+		rootSync := fake.RootSyncObjectV1Beta1(configsync.RootSyncName)
+		err := nt.Delete(rootSync, client.PropagationPolicy(metav1.DeletePropagationForeground))
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				nt.T.Error(err)
+			}
+		}
+
+		// Delete Deployment 1
+		deployment1 := fake.DeploymentObject(core.Name(deployment1Name), core.Namespace(testNs))
+		err = nt.Delete(deployment1, client.PropagationPolicy(metav1.DeletePropagationForeground))
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				nt.T.Error(err)
+			}
+		}
+
+		// Delete Namespace 1
+		namespace1 := fake.NamespaceObject(testNs)
+		err = nt.Delete(namespace1, client.PropagationPolicy(metav1.DeletePropagationForeground))
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				nt.T.Error(err)
+			}
+		}
```

Comment:
Yes, but I like cleaning them up explicitly in the right order. it makes it easier to debug when it errors. Otherwise we have to go chase down the conditions on the namespace to tell if the garbage collector failed.